src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -223,5 +223,378 @@ function (indicator_max::IndicatorMax)(u::AbstractArray{<:Any,4},
   return alpha
 end
 
+# this method is used when the indicator is constructed as for shock-capturing volume integrals
+function create_cache(::Type{IndicatorANN}, equations::AbstractEquations{2}, basis::LobattoLegendreBasis, indicator_type)
+
+  alpha = Vector{real(basis)}()
+  alpha_tmp = similar(alpha)
+  A = Array{real(basis), ndims(equations)}
+
+  if indicator_type == "NNPP"
```

Comment:
@ranocha Since this means that the actual type of the cache is now dependent on a runtime value, the point at which everything becomes a compile-time constant is moved further back. Do you see an issue with this (at the moment, I don't, since the cache is going to be of fixed type before it becomes part of the `semi`, IIRC.

Reply:
Currently, the constructor of our semidiscretization is type stable, e.g.,
```julia
julia> using Trixi

julia> equations = CompressibleEulerEquations1D(1.4);

julia> mesh = TreeMesh((-1.0,), (1.0,), initial_refinement_level=6, n_cells_max=10^5);

julia> solver = DGSEM(3, flux_lax_friedrichs);

julia> @code_warntype SemidiscretizationHyperbolic(mesh, equations, initial_condition_density_wave, solver)
Variables
  #self#::Type{SemidiscretizationHyperbolic}
  mesh::TreeMesh{1, Trixi.SerialTree{1}}
  equations::CompressibleEulerEquations1D{Float64}
  initial_condition::Core.Const(Trixi.initial_condition_density_wave)
  solver::DGSEM{LobattoLegendreBasis{Float64, 4, SVector{4, Float64}, Matrix{Float64}, Matrix{Float64}, Matrix{Float64}}, Trixi.LobattoLegendreMortarL2{Float64, 4, Matrix{Float64}, Matrix{Float64}}, SurfaceIntegralWeakForm{FluxLaxFriedrichs{typeof(max_abs_speed_naive)}}, VolumeIntegralWeakForm}
  source_terms::Nothing
  boundary_conditions::Trixi.BoundaryConditionPeriodic
  RealT::Type{Float64}
  uEltype::Type{Float64}
  initial_cache::NamedTuple{(), Tuple{}}

Body::SemidiscretizationHyperbolic{TreeMesh{1, Trixi.SerialTree{1}}, CompressibleEulerEquations1D{Float64}, typeof(initial_condition_density_wave), Trixi.BoundaryConditionPeriodic, Nothing, DGSEM{LobattoLegendreBasis{Float64, 4, SVector{4, Float64}, Matrix{Float64}, Matrix{Float64}, Matrix{Float64}}, Trixi.LobattoLegendreMortarL2{Float64, 4, Matrix{Float64}, Matrix{Float64}}, SurfaceIntegralWeakForm{FluxLaxFriedrichs{typeof(max_abs_speed_naive)}}, VolumeIntegralWeakForm}, NamedTuple{(:elements, :interfaces, :boundaries), Tuple{Trixi.ElementContainer1D{Float64, Float64}, Trixi.InterfaceContainer1D{Float64}, Trixi.BoundaryContainer1D{Float64, Float64}}}}
1 ─ %1  = Trixi.nothing::Core.Const(nothing)
│         (source_terms = %1)
│   %3  = Trixi.boundary_condition_periodic::Core.Const(boundary_condition_periodic)
│         (boundary_conditions = %3)
│   %5  = Trixi.real(solver)::Core.Const(Float64)
│         (RealT = %5)
│   %7  = RealT::Core.Const(Float64)::Core.Const(Float64)
│         (uEltype = %7)
│   %9  = Trixi.NamedTuple()::Core.Const(NamedTuple())
│         (initial_cache = %9)
│   %11 = Trixi.:(var"#SemidiscretizationHyperbolic#629")(source_terms, boundary_conditions, RealT::Core.Const(Float64), uEltype::Core.Const(Float64), initial_cache, #self#, mesh, equations, initial_condition, solver)::SemidiscretizationHyperbolic{TreeMesh{1, Trixi.SerialTree{1}}, CompressibleEulerEquations1D{Float64}, typeof(initial_condition_density_wave), Trixi.BoundaryConditionPeriodic, Nothing, DGSEM{LobattoLegendreBasis{Float64, 4, SVector{4, Float64}, Matrix{Float64}, Matrix{Float64}, Matrix{Float64}}, Trixi.LobattoLegendreMortarL2{Float64, 4, Matrix{Float64}, Matrix{Float64}}, SurfaceIntegralWeakForm{FluxLaxFriedrichs{typeof(max_abs_speed_naive)}}, VolumeIntegralWeakForm}, NamedTuple{(:elements, :interfaces, :boundaries), Tuple{Trixi.ElementContainer1D{Float64, Float64}, Trixi.InterfaceContainer1D{Float64}, Trixi.BoundaryContainer1D{Float64, Float64}}}}
└──       return %11
```
Adding this check here means that we're making it significantly more difficult for the compiler, which can increase latency and will disallow any improvements from better precompilation etc. Moreover, the same check is also used at runtime, e.g., https://github.com/trixi-framework/Trixi.jl/blob/cd8ba83a99cde5e0d353ec7a12a37c32b617a72f/src/solvers/dgsem_tree/indicators_2d.jl#L266. This increases code bloat significantly, makes the life of the compiler much more difficult, and can reduce performance. Why do you not use some singletons to encode the `indicator_type` instead of a string? Then, you can use dispatch to make everything type stable.

Reply:
Additionally using hard-coded `if` checks like this makes the code 
- hard to read (everything is bundled in one place)
- not extensible (which you can from the outside if relying on dispatch)

Reply:
Singletons as type parameters, as in, e.g., `IndicatorANN{NNPP}` with `struct NNPP end` and then using `if indicator_type(ind) == NNPP() ...`?

Reply:
I acknowledge that the `if` is not optimal, but did I understand the singleton stuff right as you intended?

Reply:
Singleton type such as https://github.com/trixi-framework/Trixi.jl/blob/83bd625bc335c75ed99a905f31d022231ca59c8c/src/basic_types.jl#L62-L69. For example,
```julia
julia> struct NeuralNetworkIndicatorPP end

julia> const NNPP = NeuralNetworkIndicatorPP()
NeuralNetworkIndicatorPP()

julia> # NNPP gets exported

julia> struct IndicatorANN{IndicatorType}
           indicator_type::IndicatorType
           # other stuff
       end

julia> indicator = IndicatorANN(NNPP)
IndicatorANN{NeuralNetworkIndicatorPP}(NeuralNetworkIndicatorPP())

julia> function now_im_dispatching_on_the_indicator_type(indicator::IndicatorANN{typeof(NNPP)})
           println("I'm type stable and running with NNPP.")
       end
now_im_dispatching_on_the_indicator_type (generic function with 1 method)

julia> now_im_dispatching_on_the_indicator_type(indicator)
I'm type stable and running with NNPP.
```

Reply:
Don't use these `if` checks inside functions for basic stuff like this. This is not extensible and bloats up the code, increasing latency and possibly reducing runtime performance since the compiler might need to give up earlier.

Reply:
> I acknowledge that the `if` is not optimal, but did I understand the singleton stuff right as you intended?

Yes